### PR TITLE
core: improve clist

### PR DESCRIPTION
**Pull request by kaspar030**
After working with redis for a while, I now find it's naming for list operations way more intuitive than what we had before, so I adapted our clist implementation accordingly.

While at it, I added some functions and (IMHO) improved the docs quite a bit.

*edit*

With the improvements from this PR, clist becomes a generally usable data structure that can be used wherever a list, stack (lifo) or queue (fifo) is needed, with O(1) operations for adding objects at head or tail and removing objects from head, in addition to O(n) operations for remove from tail, remove any object and find any object.

### CODE REVIEW

**Comment by gebart** [comment]
Could you add an implementation for `clist_print` as well while you're working on clist?

**Comment by gebart** [comment]
I like the naming changes, it makes the API clearer.
I have tested some of the changes in the unit tests for VFS in #5616 (`clist_find`, `clist_remove`, `clist_rpush`)

File: core/include/clist.h [hunk 1]
```diff
@@ -14,20 +14,64 @@
...
+ *
+ * To use clist as a queue, use clist_rpush() for adding elements and clist_lpop()
+ * for removal. Using clist_lpush() and clist_rpop() is inefficient due to
+ * clist_rpop()'s O(n) runtime.
+ *
+ * To use clist as stack, use clist_lpush()/clist_lpop().
+ *
+ * Implementation details:
  *
  * Each list is represented as a "clist_node_t". It's only member, the "next"
```
Comment (gebart) [inline core/include/clist.h]: "It's" -> "Its"

File: core/include/clist.h [hunk 2]
```diff
@@ -14,20 +14,64 @@
...
+ * the container_of() macro in list operations.
+ * See @ref thread_add_to_list() as example.
  *
  * @author      Kaspar Schleiser <kaspar@schleiser.de>
  */
 
 #ifndef CLIST_H
 #define CLIST_H
 
+#include <stdio.h>
```
Comment (gebart) [inline core/include/clist.h]: not used anywhere as far as I can see?

**Comment by gebart** [comment]
Can you improve the unit test coverage? `clist_lpush` and `clist_rpop` are not tested currently

File: core/include/clist.h (hunk 2 above)
Comment (kaspar030) [inline core/include/clist.h]: yup, addressed.

File: core/include/clist.h (hunk 1 above)
Comment (kaspar030) [inline core/include/clist.h]: addressed

**Comment by gebart** [comment]
(https://github.com/RIOT-OS/RIOT/pull/5616#issuecomment-233168950)
@kaspar030 wrote:
>you not feeling confident about it's usage makes me wonder if the API is intuitive like this?

I don't think there's anything wrong with the API in this PR, rather the documentation is missing some pieces. The missing piece for me was the fact that the list is represented by a `clist_node_t ` item which is outside of the list circle but pointing in (`_vfs_mounts_list` in #5616).

Another thing is the iteration of lists. The example could be made more concrete, I'll try to come up with some improvements and post back here.
The name of the function `clist_advance` made me think that I should use that for iteration. Maybe the example code should use that?

I think this does the same thing but with `clist_advance` instead of manually advancing, with the downside that the `container_of` macro needs to be passed the `node->next` pointer instead of just `node` (untested code)
```c
/* outside of function */
struct {
    int value1;
    char str[8];
    clist_node_t list_entry;
} myitem_t;

/* List handle */
clist_node_t mylist;

/* In function where we need to iterate over the list once: */

/* Make a copy of the list handle to use for iteration */
clist_node_t node = mylist;

if (node->next == NULL) {
    /* list empty */
    /* do some special handling */
    return;
}
do {
    /* advance the iteration handle to point at the next item */
    clist_advance(node);
    /* find a myitem_t pointer given a pointer to item->list_entry */
    myitem_t *item = container_of(node->next, myitem_t, list_entry);
    /* do something with *item */
    /* example: */
    if (strcmp(item->str, "hej") == 0) {
        return item;
    }
    /* ... */
    /* check if we have gone round the list to the last item */
} while (node.next != _vfs_mounts_list.next);
```

**Comment by gebart** [comment]
Let's see what Murdock says.

File: core/include/clist.h (hunk 1 above)
Comment (gebart) [inline core/include/clist.h]: did you forget to git push?

**Comment by kaspar030** [comment]
- updated unittests
- fixed headers

I've added a "clist_foreach()" function that can be used for traversal using a function pointer (and added an example on how to use it to print clist entries).

How about renaming "clist_advance()" to "clist_rotate()", "clist_rotate_clockwise()" or "clist_rotate_right()"?

Would it make sense to typedef "clist_t" to "clist_node_t" and then use the former as list reference object?

File: core/include/clist.h [hunk 3]
```diff
@@ -40,23 +93,19 @@
...
- * the container_of() macro in list operations.
- * See @ref thread_add_to_list() as example.
  */
 typedef list_node_t clist_node_t;
 
 /**
- * @brief inserts *new_node* into *list*
+ * @brief Appends *new_node* at the end of *list
+ *
+ * @note Complexity: O(1)
```
Comment (miri64) [inline core/include/clist.h]: Matter of taste, but surrounding `$` to change to LaTeX parsing?

Comment (miri64) [inline core/include/clist.h]: (to be in line with other, more complex mathematical terms)

**Comment by gebart** [comment]
After this is merged, does it make sense to deprecate the other `list.h` and update the users to use `clist.h` instead?
`list_remove_head` should be replaceable by `clist_lpop`, if I am not mistaken.
Current users of `list_remove_head` are `mutex.c` and `msg.c`.
`list_add` is not used anywhere in the main repository, as far as I can see.

File: core/include/clist.h (hunk 3 above)
Comment (gebart) [inline core/include/clist.h]: if you want to use LaTeX parsing for the complexity you might as well go all in and write it `$\mathcal{O}(n)$`

see http://texblog.org/2014/06/24/big-o-and-related-notations-in-latex/ for an example

**Comment by gebart** [comment]
Why is not `clist_foreach` inlined when everything else is?
The `clist_foreach` function does not look very much more complex than anything else in the `clist.h` header and it would likely give the compiler an opportunity for some optimizations for the case when the function pointer is a compile time constant.

**Comment by miri64** [comment]
> After this is merged, does it make sense to deprecate the other list.h and update the users to use clist.h instead?

So back to how it used to be ^^?

**Comment by gebart** [comment]
I don't think clist_rotate_right/clockwise are more speaking than clist_advance. In fact, I find it hard to understand which part is moving clockwise, depending on how you picture the list.

If you picture the circular list items as the face of a clock and the list handle is the hand of the clock, then it makes sense to call it rotate clockwise, but on the other hand, if you picture the list handle as fixed and the circular list as the moving part (wheel of fortune style), then rotate clockwise is the opposite operation.

I have never worked with Redis but I looked up the manual after I read your description on this PR. There is an operation called RPOPLPUSH which does the same thing, except it takes two arguments, see http://redis.io/commands/rpoplpush. Maybe that is something to consider for the `clist_advance` functionality?

**Comment by gebart** [comment]
@miri64 you are right (https://github.com/RIOT-OS/RIOT/pull/4557 introduced the other list.h). I see no reason for list.h to live on after this is merged.

**Comment by kaspar030** [comment]
> After this is merged, does it make sense to deprecate the other list.h and update the users to use clist.h instead?

I thought about it, but then again, clist has (small) overhead over list.h, and it's main users are msg and mutex, where that kind of overhead is actually visible.

File: core/include/clist.h (hunk 3 above)
Comment (kaspar030) [inline core/include/clist.h]: If you don't insist, I would keep it ascii, as some people do look at the code to read the documentation, and in that case ```O(n)``` is a lot more pleasant to the eyes than ```thcal{O}(n)$```.

**Comment by kaspar030** [comment]
> RPOPLPUSH which does the same thing, except it takes two arguments

Hm. That operation is supposed to be used for reliable queueing (rpop from one list, lpush to another, atomically), and that it can be used on the same list is a side effect. But then it does what clist_advance() does (or better, lpoprpush would do). I think I like it better than _advance(), but I'm a little afraid that people forget this is still a circular list. :)

**Comment by kaspar030** [comment]
> So back to how it used to be ^^?

mutex and msg used priority_queue before.

**Comment by kaspar030** [comment]
- rebased
- renamed ```clist_advance()``` -> ```clist_lpoprpush()```
- added ```clist_lpeek()``` and ```clist_rpeek()```

**Comment by gebart** [comment]
@kaspar030 
>*edit* I'll do some more benchmarking, but let's keep list.h for now (this PR), ok?

Yes, let's leave the list.h change for a separate PR.

File: core/include/clist.h (hunk 3 above)
Comment (miri64) [inline core/include/clist.h]: $O(1)$

**Comment by kaspar030** [comment]
I think this should be mergable, so I'll squash. @gebart @miri64 do you feel like ACK today? ;)

**Comment by gebart** [comment]
Please add a minimal unit test for `clist_foreach`.
ACK from me once that is done. Needs a second ACK as well

File: core/include/clist.h [hunk 4]
```diff
@@ -98,23 +171,175 @@ static inline clist_node_t *clist_remove_head(clist_node_t *list)
...
+ * @note Complexity: O(n)
+ *
+ * @param[in]       list    pointer to clist
+ * @param[in,out]   node    Node to remove for
+ *                          Must not be NULL.
+ *
+ * @returns         node if found and removed
+ * @returns         NULL if node is not a list member
+ */
+static inline clist_node_t *clist_remove(clist_node_t *list, clist_node_t *node)
```
Comment (miri64) [inline core/include/clist.h]: I think this function is big enough to be a non-inline function.

File: tests/unittests/tests-core/tests-core-clist.c [hunk 5]
```diff
@@ -22,84 +22,198 @@ static list_node_t test_clist;
...
+    TEST_ASSERT_NOT_NULL(list->next);
+
+    TEST_ASSERT(clist_lpop(list) == &tests_clist_buf[2]);
+    TEST_ASSERT_NULL(list->next);
+    TEST_ASSERT_NULL(clist_lpop(list));
+}
+
+static void test_clist_lpush(void)
+{
+    set_up();
```
Comment (miri64) [inline tests/unittests/tests-core/tests-core-clist.c]: set_up() is called automatically before this function is called so why do you call it?

File: core/include/clist.h [hunk 6]
```diff
@@ -98,23 +171,175 @@ static inline clist_node_t *clist_remove_head(clist_node_t *list)
...
+/**
+ * @brief Traverse clist, call function for each member
+ *
+ * If @p func returns non-zero, traversal will be aborted like when calling
+ * break within a for loop.
+ *
+ * @param[in]       list        List to traverse.
+ * @param[in]       func        Function to call for each member.
+ */
+static inline void clist_foreach(clist_node_t *list, int(*func)(clist_node_t *))
```
Comment (miri64) [inline core/include/clist.h]: This one too, maybe.

Comment (gebart) [inline core/include/clist.h]: I wrote in an earlier comment that this function should be inlined to allow the compiler to optimize for the case where `func` is compile time constant, which is likely the most common case.
(the comment was deleted by github when the file no longer exists)

File: core/include/clist.h (hunk 4 above)
Comment (kaspar030) [inline core/include/clist.h]: Some of them might be. I kept them static inline as they're used by the scheduler, where every instruction counts. I'll do an evaluation of the benefits/downsides in a folo.

**Comment by kaspar030** [comment]
- added clist_foreach() unittest, immediately squashed

**Comment by kaspar030** [comment]
hmm, murdock is not setting the new status.

File: core/include/clist.h (hunk 6 above)
Comment (miri64) [inline core/include/clist.h]: Okay, sorry for the noise then.

File: core/include/clist.h (hunk 4 above)
Comment (miri64) [inline core/include/clist.h]: kk.

File: tests/unittests/tests-core/tests-core-clist.c [hunk 7]
```diff
@@ -70,36 +173,107 @@ static void test_clist_remove_two(void)
...
+        return 1;
+    }
+    else {
+        return 0;
+    }
+}
+
+static void test_clist_foreach(void)
+{
+    set_up();
```
Comment (miri64) [inline tests/unittests/tests-core/tests-core-clist.c]: - set_up() ;-)

File: tests/unittests/tests-core/tests-core-clist.c [hunk 8]
```diff
@@ -22,46 +22,149 @@ static list_node_t test_clist;
...
-    clist_insert(list, elem);
+    clist_rpush(list, elem);
 
     TEST_ASSERT_NOT_NULL(list->next);
     TEST_ASSERT(list->next->next == list->next);
 }
 
 static void test_clist_add_two(void)
 {
+    set_up();
```
Comment (miri64) [inline tests/unittests/tests-core/tests-core-clist.c]: \- set_up()

File: tests/unittests/tests-core/tests-core-clist.c [hunk 9]
```diff
@@ -22,46 +22,149 @@ static list_node_t test_clist;
 static void set_up(void)
 {
     memset(tests_clist_buf, 0, sizeof(tests_clist_buf));
+    test_clist.next = NULL;
 }
 
-static void test_clist_add_one(void)
+static void test_clist_rpush(void)
 {
+    set_up();
```
Comment (miri64) [inline tests/unittests/tests-core/tests-core-clist.c]: \- set_up()

File: tests/unittests/tests-core/tests-core-clist.c (hunk 8 above)
Comment (kaspar030) [inline tests/unittests/tests-core/tests-core-clist.c]: nope, this function is called from the other tests, and they expect a defined state.

File: tests/unittests/tests-core/tests-core-clist.c (hunk 7 above)
Comment (kaspar030) [inline tests/unittests/tests-core/tests-core-clist.c]: clist needs to be clear before the test starts, but this test doesn't call one of the _add_n() functions.

File: tests/unittests/tests-core/tests-core-clist.c (hunk 8 above)
Comment (miri64) [inline tests/unittests/tests-core/tests-core-clist.c]: Ah, kk. Can you please mention this in a comment?

File: tests/unittests/tests-core/tests-core-clist.c (hunk 7 above)
Comment (miri64) [inline tests/unittests/tests-core/tests-core-clist.c]: But `set_up()` is already run by the test system before the call to this function to ensure that.

**Comment by kaspar030** [comment]
@miri64 Didn't know set_up() was called.

- removed redundant set_up() calls
- immediately squashed

**Comment by kaspar030** [comment]
Again the status doesn't update. The build was successful (https://ci.riot-labs.de/RIOT-OS/RIOT/5629/a3547963dc4afe001dc37599304c60f784fd2b3b/output.html).

**Comment by kaspar030** [comment]
Now it's green. :)

**Comment by miri64** [comment]
ACK and go.

**Comment by basilfx** [comment]
@kaspar030 I get a build error due to missing return in this method (using Clang): https://github.com/RIOT-OS/RIOT/blob/master/core/include/clist.h#L195

The `return NULL;` is missing.

**Comment by kaspar030** [comment]
@basilfx yup, I PR'ed a fix.

**Pull request merged**